src/DataSet/ObjectDataSet.php

```diff
@@ -9,43 +9,114 @@
 use ReflectionProperty;
 use Yiisoft\Validator\AttributeEventInterface;
 use Yiisoft\Validator\DataSetInterface;
+use Yiisoft\Validator\ObjectDataSetInterface;
 use Yiisoft\Validator\RuleInterface;
 use Yiisoft\Validator\RulesProviderInterface;
 
 use function array_key_exists;
 
 /**
  * This data set makes use of attributes introduced in PHP 8. It simplifies rules configuration process, especially for
- * nested data and relations. Please refer to the guide for example.
+ * nested data and relations. Please refer to the guide for examples.
  *
  * @link https://www.php.net/manual/en/language.attributes.overview.php
  */
-final class ObjectDataSet implements RulesProviderInterface, DataSetInterface
+final class ObjectDataSet implements RulesProviderInterface, ObjectDataSetInterface
 {
     private bool $dataSetProvided;
-
     /**
      * @var ReflectionProperty[]
      */
-    private array $reflectionProperties = [];
-
-    private iterable $rules;
+    private ?array $reflectionProperties = null;
+    private ?iterable $rules = null;
+    private ?array $data = null;
 
     public function __construct(
         private object $object,
-        private int $propertyVisibility = ReflectionProperty::IS_PRIVATE|ReflectionProperty::IS_PROTECTED|ReflectionProperty::IS_PUBLIC
+        private int $propertyVisibility = ReflectionProperty::IS_PRIVATE |
+        ReflectionProperty::IS_PROTECTED |
+        ReflectionProperty::IS_PUBLIC
     ) {
-        $this->parseObject();
+        $this->dataSetProvided = $this->object instanceof DataSetInterface;
     }
 
-    public function getObject(): object
+    public function getRules(): iterable
     {
-        return $this->object;
+        if ($this->rules !== null) {
+            return $this->rules;
+        }
+
+        $objectHasRules = $this->object instanceof RulesProviderInterface;
+        $rules = $objectHasRules ? $this->object->getRules() : [];
```

Comment:
@vjik's suggestion: Do not perform caching via static variable when object has its own `getRules()` implementation.